### ISSUE TITLE
`interaction` tests using `.hs` script are failing on github actions

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#5258]
The `interaction` testsuite started failing on github, see e.g. https://github.com/agda/agda/runs/2028943853?check_suite_focus=true, producing errors like:
```
Issue983.hs:«line»:«col»: error:
Could not load module ‘Agda.Version’
It is a member of the package ‘Agda-2.6.2-F2BRXV7od3W7ndov7ntyY4’
which is unusable due to missing dependencies:
aeson-1.5.5.1-7cTfqertSdS3IIyN4VfUci ...
...
6 | import Agda.Version
| ^^^^^^^^^^^^^^^^^^^
```
I have seen this with another test case which also as a `.hs` script (cf. 9bb868bc460af564b02a6e4b29e8182dc8824519).

**Comment by andreasabel** [comment]
Actually also failing locally on my machine (I am using GHC-9.0.1).

Could be a `.ghc.environment` thing? (Wild guess.)  Libraries are not installed globally anymore...

**Comment by asr** [comment]
I don't understand all the details but it is a consequence of mixing Cabal v1- and v2-commands for installing different programs. I only use Cabal v1-commands because Agda uses them, when I install something using v2-commands I have a similar problem.

**Comment by UlfNorell** [comment]
Yes, I'm pretty sure this is a ghc environment problem. I had the same problem and fixed it by finding and nuking all environment files (I forget how to find where they are), and putting `write-ghc-environment-files: never` in my cabal config.

**Comment by andreasabel** [comment]
> nuking all environment files

Are you installing everything with `v1`-cabal?
Otherwise, can you still use `ghci` then?  It doesn't pick up packages otherwise, it seems. 

So far, `write-ghc-environment-files: always` worked well for me. I followed the advice at https://github.com/sol/doctest/issues/245#issuecomment-596038515.

**Comment by UlfNorell** [comment]
I'm using a mix of v1 and v2, and I haven't had any problems since I got rid of the environment nonsense.

**Comment by andreasabel** [comment]
On my machine, the problem seems to be this:
- I do a `cabal v2-build` which creates a `.ghc.environment` file
- after a while, I do `make install-bin` which installs a new version of Agda
- the `.ghc.environment` file is now stale; the hashes appended to the package-ids are wrong
- `make interaction` now fails, as `runghc` picks up the stale environment

Maybe the interaction suite could be made more robust by passing the `package-env` option: https://ghc.gitlab.haskell.org/ghc/doc/users_guide/packages.html#ghc-flag--package-env%20%E2%9F%A8file%E2%9F%A9|%E2%9F%A8name%E2%9F%A9

UPDATE: After removing the `.ghc.env*` files, `ghci` called from my `haskell-mode` in emacs does not find the modules listed in the `.cabal` file.  They are communicated by writing the `.ghc.env*` file upon `cabal v2-build`.  A workaround is the following:
1. make a `.ghci` file in the `src/` folder of the project 
2. add `:set -package foo` there for each `build-dependency` `foo` in the `.cabal` file.

Item 1. is already necessary in my case to get the `LANGUAGE` flags from the `default-extensions`.

**Comment by andreasabel** [comment]
The testsuite uses `stack` to build and run Agda: https://github.com/agda/agda/blob/c28ec33cc563970f26ba3f62e24f0179ef1a4af9/.github/workflows/test.yml
I think I don't know `stack` enough to fix this.  
On my machine at least `stack build` does not generate any `.ghc.env*` files.

**Comment by andreasabel** [comment]
Looks like @L-TChen 's patch 8d2c172cc2bc700e6eed993557ae11b19cedf682 has fixed the problem: https://github.com/agda/agda/actions/runs/631557577

**Comment by L-TChen** [comment]
@andreasabel it is only a temporary fix. 

The problem is that `stack` calculates the hash of dependencies and *something else* which changes from time to time, even though the generated lock file `stack-x.y.z.yaml.lock` is the same. However, as the GitHub workflow `test.yml` caches the global store of `stack` using the hash of `stack-x.y.z.yaml.lock` as key only, `stack` apparently creates a new global store from time to time. As a result, the job `cubical-benchmark-misc` of the workflow `test.yml` does not have the right global store to run `interaction` tests even the cache is hit. 

Therefore, to use the new global store, a temporary fix is to use a new cache key (e.g., a newer LTS version) every time the hash is changed. I will keep this open until there is a proper fix.

**Comment by jespercockx** [comment]
Since this is unrelated to the release of 2.6.2, I've changed the milestone to `icebox`.

**Comment by andreasabel** [comment]
The ghost is back to haunt us: https://github.com/agda/agda/runs/2408325047?check_suite_focus=true

**Comment by L-TChen** [comment]
A-ha, at least I think I know why. Oddly enough, stack does not generate a different lock file from the previous one without including the locally patched `text-icu` so the cache is hit (which it shouldn't). 

It is also my bad that I removed the hash of `stack.yaml` as part of cache key by mistake.

**Comment by andreasabel** [comment]
Seems like the recipe to fix these problems is to reset the CI cache.

**Comment by andreasabel** [comment]
@plt-amy : You got hit by this in https://github.com/agda/agda/actions/runs/3526968162.
Actually, you pointed this out in:
- https://github.com/agda/agda/pull/6313#issuecomment-1324220969

I didn't make the connection at that time, but I fixed this fluke in 9d4e34313208b56618161c5d96839fc0637731f1 and https://github.com/agda/agda/commit/cf8947bd3f27958ec2547c2cbe525d7fea8ba3da (blundered it first).

We haven't solved this problem systematically, but the workaround it to bump the cache key in the `test.yml` workflow.
To not shoot myself in the foot all the time when doing this, I opened:
- https://github.com/agda/agda/pull/6358
